### DISCUSSION TITLE
Changing patches headless with MIDI?

### DISCUSSION THREAD

**Discussion by Stooovie** [probonopd/MiniDexed#496] [General]
Hi, I'm a total N00b, goal is to just plug my Nektar LX88 keyboard and play, sort of what I can do with FluidPi/Squishbox running headless.

Is there a way to just be able to change patches with MIDI CCs (preferably 112/111), and/or knobs? The Nektar sends MIDI at channel 16.

I did look around trying to find out but I failed, you guys are way too far ahead :) Any tips? Thanks!

**Comment by diyelectromusic** [comment]
There are a few options really, but nothing exactly what you're asking for:
1) Use "MIDI buttons" to set up certain MIDI note or CC events to control the UI of MiniDexed, but that will require a display to be able to see where you are in the menu structure...
2) Use MIDI Program Change messages.  There is an option to recognise programs 1 through 128 and map these onto the first four banks.
3) Use MIDI BANKSEL and Program Change messages to select any of 16384 banks installed and then any of the 1 to 32 voices within them.

We don't currently have a simple "voice up/down" MIDI message, but you might be able to program your controller to output some combination of BANKSEL and Program Change.  I had a quick look at the spec, but I must admit I didn't spot how to do that directly, but it looks very configurable!

Kevin

**Comment by probonopd** [comment]
Hi @Stooovie, welcome. Great point.
When you are saying "patch", this could mean two things:
* __Voice:__ one voice is what runs on one Tone Generator (TG) (a.k.a. "one DX7 instance")
* __Performance__: up to 8 voices plus effects (a.k.a. "similar to a TX802, a.k.a 8 DX7 modules plus effects")

Currently we are switching __voices__ upon MIDI Program Change messages, based on which bank (cartridge) is loaded via MIDI Bank Select (MSB+LSB).

I assume what you are asking is to switch __performances__ in a similar way.
I agree that this would be a very useful capability.

It is still not entirely clear to me how this is handled by "real" instruments like the TX802.

The TX812 is essentially a collection of TX7-like modules with no concept of performances as far as I know. That is how MiniDexed originally started out.

Later we added performances, vaguely similar to the TX802.

Looking at the TX802 Owners Manual on page 32, there is a diagram that suggests one can change both voices and performances via MIDI. Page 31 mentions a "prgram change sw". It seems that depending on whether the user has pressed the "PROGRAM SELECT" (Performance Mode) or the "VOICE SELECT" (Edit Mode) phyiscal button on the device, the same MIDI CC commands either switch programs or voices.

But is that good? It means that you have to select this on the device itself and can't seem to switch these modes via MIDI.

Or are there SysEx messages to switch between Performance Mode and Edit Mode?

How do other instruments handle this?
Are there any kind of conventions?

If we cannot find out how other instruments do this, then I could imagine to do the following:
* Add a menu "MIDI CC Mode" containing "Performance" and "Voice"
* When set to "Performance", MIDI Program Change messages change performances
* When set to  "Voice",  MIDI Program Change messages change voices (as it is now). The setting should be persisted in `minidexed.ini`
* We introduce (define) a MiniDexed proprietary MIDI SYSEX for switching between "Performance" and "Voice" (for those who want to do the switching without menu diving)
* While we are at it, we might want to consider MiniDexed proprietary MIDI SYSEX for just about everything else in `minidexed.ini`

**Reply by probonopd** [reply]
This is what ChatGPT thinks how it works on the DX7IIfd (still needs to be verified with proper sources):

>  It is not possible to switch between voice mode and performance mode on the Yamaha DX7IIfd using MIDI messages alone. The mode selection between voice mode and performance mode is a local setting on the synthesizer itself and cannot be changed via MIDI commands.
>
> To switch between voice mode and performance mode on the DX7IIfd, you need to use the front panel controls on the synthesizer. Typically, there is a dedicated button or menu option on the DX7IIfd that allows you to toggle between voice mode and performance mode.
>
> Once you have selected the desired mode using the front panel controls, MIDI messages will interact with the DX7IIfd based on the current mode setting. **If the synthesizer is set to voice mode, MIDI Program Change messages will change individual voices within the current performance. If the synthesizer is set to performance mode, MIDI Program Change messages will switch to entirely new performances.**

But as so often, ChatGPT changes its mind when challenged:

> Apologies for the confusion in my previous response. You are correct; there is a MIDI System Exclusive (SYSEX) message that can be used to change the mode of the Yamaha DX7IIfd between voice and performance modes.
> 
> The SYSEX message for mode change on the DX7IIfd is as follows:
> 
> F0 43 1n 4C 00 F7
> 
> Here's how you can interpret and use this SYSEX message:
> 
> F0: Start of the SYSEX message.
43: Yamaha manufacturer ID.
1n: The MIDI channel (where n represents the desired channel number in hexadecimal, ranging from 0 to F).
4C: The command for mode change.
00: The data value for the mode. Use 00 for voice mode and 01 for performance mode.
F7: End of the SYSEX message.
To switch the DX7IIfd to voice mode using MIDI SYSEX, you would send the following message on the desired MIDI channel (replace n with the hexadecimal representation of the channel number):
> 
> F0 43 1n 4C 00 F7
> 
> To switch the DX7IIfd to performance mode, you would send the following message:
> 
> F0 43 1n 4C 01 F7

Is it true though?

**Reply by diyelectromusic** [reply]
I'm afraid I switch off whenever I see a "this is what ChatGPT says..." type post.

The problem is that we need an expert on the DX7 to tell us if ChatGPT is telling us accurately, in which case why not just ask the expert directly?  Especially considering that ChatGPT would only "know" the answer from leeching off the experts who chose to document this in the first place (with no attribution)...

It's your project of course, but I think this kind of response is just adding noise to the conversation - instead of attempting to find out directly, we're now down the rabbit hole of attempting to work out if this is right... and what if it is 95% right, but we don't spot the 5% it has made up?  We just can't trust its output for technical things in my view unless we already have the expertise to know it has made something up...

And that is not even getting into the whole circular reference thing... now this is text on GitHub, how long before this gets fed back into the machine, thus reinforcing the strength of the unverified output?

Kevin

**Reply by probonopd** [reply]
Yes, you are probably right. I find the tool useful for some situations in which you can immediately verify its answers, but if you can't (e.g., due to to the lack of corresponding hardware in my case), it's not really that useful. Hence my clear (dis)qualifiers ("what ChatGPT thinks", "still needs to be verified with proper sources", "Is it true though?").

**Reply by diyelectromusic** [reply]
I think this is the key worry for me.  If this had given you something that worked, then we'd have probably just said "Score 1 for ChatGPT" and stuck it in.

But we probably wouldn't have noticed that the 1n is actually the Yamaha Device number, not MIDI channel.  For us that is the TG number.  With many default setups we'd not notice an issue, but somewhere further down the line we'd end up with some obscure bug because we'd assumed device and channel were the same thing...

This is why it feels really irresponsible for them to have released this.  It is designed to look like it is giving authoritative, confident answers, when actually it only gives things that look like authoritative, confident answers, but won't be.

And that is before you get me started on how it exploits human knowledge to do it. I write my blog and release my own code to help other people, not other machines. Of course one may argue that surfacing my material through a machine is helpful, and it is, its called search, and the quid pro quo is that it directs you back to the source, but the key context that is being broken here is that the machine isn't pointing you at my information - it is presenting a mashed up version of it as its own in a very confincing way.  In my view that is ethically and morally wrong and devalues the human intent behind providing the information in the first place.

And my apologies to @Stooovie for going so far off topic - I suspect you are now wondering what on earth you've kicked off with your otherwise very relevant question!!

Don't worry - normal service will be resumed shortly :)

Kevin

**Reply by Stooovie** [reply]
@diyelectromusic no worries at all! I love the project BUT I'm also not a programmer nor a MIDI connosseur and I would just love to plug the keyboard into the Pi and play a couple of readymade patches, if possible at all :)

**Reply by diyelectromusic** [reply]
Well if someone can verify that we can enable performance mode from MIDI using the sequence I posted below, then I'll have a look at allowing program change to be either performance or voices.

In fact, I can look at it regardless and make it a UI/config setting, but I'd still like to know if it was possible over MIDI! :)

But I don't know if what you're asking is for voice changing within a single TG (which we can do as per my first post) or "performance" changing for the whole multi-TG instrument (which we can't yet).

Kevin

**Comment by diyelectromusic** [comment]
This is from the DX7II FD/D Advanced MIDI Applications manual... but not having a real DX7 (any version), I'm not sure if this helps in the context of the question or not...

![image](https://github.com/probonopd/MiniDexed/assets/68612569/42bc6b83-562c-4cbc-8945-253024f0fe35)

In the MIDI Technical Data and Charts supplement, under Program Change, it just lists "Program Number" but then later on in the Reception Data section is shows the following:

![image](https://github.com/probonopd/MiniDexed/assets/68612569/eba3bda3-9c88-4328-85b8-ac61bb86f001)

From the "Quick Reference Guide", we can see there is a UI setting for Performance or Program:

![image](https://github.com/probonopd/MiniDexed/assets/68612569/0c61dfda-7708-416b-844f-af1bac0a05f1)

But what I'm curious about now, is the PPCMOD entry in the "System Setup Parameters" section of the MIDI Techinical Data...

![image](https://github.com/probonopd/MiniDexed/assets/68612569/015ca4c7-a5d0-417a-bb00-bd9a0b3c03a9)

And this entry, which implies all the front panel switches have a MIDI SysEx code... if I read this correctly?

![image](https://github.com/probonopd/MiniDexed/assets/68612569/ff356ea7-460e-49b3-a22c-86dd7fca8177)

But either way this is going to take some decoding... MIDI specs for vintage synths aren't the easiest things in the world to read :)

Update: Ok, so I think it is saying to change PPCMOD:
F0 43 1n 19 4A 0m F7 where m = 0 for OFF, 1 = Normal, 2 = Programmable

So this doesn't switch between normal and performance mode, so the only option might be the "remote switch" which I think is 41 in that figure... So this might turn on performance mode:
F0 43 1n 1B 29 7F F7

And this might turn it off:
F0 43 1n 1B 29 00 F7

But confusingly the SysEx parameters are shown as binary variables, but I think all the values are given in decimal!  Hence 0x29 = 41 I think is the right switch to use...

Anyone have a DX7 II and MidiOx and want to try it!?

Another Update: Ok, so this, confusingly, implies it is 0x18 not 0x1B...

![image](https://github.com/probonopd/MiniDexed/assets/68612569/5dc85c40-660c-409e-8a91-bb6e479e32d6)


Kevin

**Comment by diyelectromusic** [comment]
I have a speculative branch that adds program up/down and tone generator up/down buttons and MIDI buttons to MiniDexed if anyone fancies giving it a go: https://github.com/diyelectromusic/MiniDexed/commits/ProgramButtons

Setting the following in the minidexed.ini file for this build would do what you were asking I think:
MIDIButtonCh=16
MIDIButtonNotes=0
MIDIButtonPgmUp=112
MIDIButtonPgmDown=111

It defaults to the Tone Generator selected by the UI (or TG1 is nothing is selected), but the TG can be changed by adding
MIDIButtonTGUp=
MIDIButtonTGDown=

A different update might be able to look at making Program Change with either voices or performances... watch this space...

Kevin

**Reply by diyelectromusic** [reply]
Latest version now allows choice of either Voices or Performances for Program Change messages...
Set PerformanceProgramChange=1 to switch to choosing Performances.

But I'm getting stuck notes when playing Performances, so don't know if my RPi 3 just isn't up to more complex sounds - any thoughts?  I'll try with the latest proper release just to be sure.

Update: Ok ignore that - seems to be an issue with my Aturia MiniLab... other USB and serial MIDI is fine.

Kevin

**Reply by diyelectromusic** [reply]
Now all included in https://github.com/probonopd/MiniDexed/pull/500

**Reply by probonopd** [reply]
Posted a build for testing at https://github.com/probonopd/MiniDexed/pull/500. Would be great if you could give it a try @Stooovie. Don't forget to edit the `minidexed.ini` file as needed.

**Reply by jayro3173740** [reply]
Hello - I'm new to the forums and minidexed. I'm using a Morningstar MC8 controller to send MIDI CC/PC messages to change patches and I've figured out how to get it to change the voices per TG via MSB/LSB. (It's a several step process which I'm happy to share if it helps)

Am I reading the above correctly that it's similar to change Performances, but you need to set PerformanceProgramChange to 1 in the minidexed menu? Or is there another way? Ideally I'd love to be able to do both things while playing live and pull up a Voice or Performance depending on which song I need it for. Is that possible or is it one or the other? Just making sure I'm understanding. Thanks!

**Reply by probonopd** [reply]
https://github.com/probonopd/MiniDexed/wiki/Files#midi says:

* `PerformanceSelectChannel` determines if MIDI Program Change (PC) messages should select individual voices within a tone generator or act at the system level and select performances.  This mirrors the similarly named control on a DX7II.  Setting this to 0 (the default) means PC messages select voices on any tone generators set to the receiving MIDI channel.  Setting this to a MIDI channel (1 to 16) means PC messages on that channel will select performances.  Setting this to a value >16 selects "Omni" mode so any PC messages received select performances.  Note when in performance mode, no PC messages will be passed on to tone generators.  Note that `MIDIRXProgramChange` has to be set to 1 for this to have any affect.

So, if you set this to a MIDI channel other than your regular MIDI channels (e.g., 1-8), then it should change performances only if you send the message to the specified MIDI channel.
